[JSC] Mark inlined JSValue methods with `inline` keyword (and fix a bunch of things that break when that happens)

jernoble opened the pull request:
#### d7deeb4d678e012b7665d53198f3d80952fffdd6
<pre>
[JSC] Mark inlined JSValue methods with `inline` keyword (and fix a bunch of things that break when that happens)
<a href="https://rdar.apple.com/164200123">rdar://164200123</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=302121">https://bugs.webkit.org/show_bug.cgi?id=302121</a>

Reviewed by Keith Miller.

Clients of JSC can face &quot;missing symbol&quot; linker errors for inlined methods when the
correct *Inlines.h headers are not included, and that task is made much, much more difficult when
the inlined methods aren&apos;t marked as &quot;inline&quot;. Adding the &quot;inline&quot; keyword causes the compiler to
throw an error when that symbol is referenced, making it much easier to notice a missing *Inlines.h
header at compile time.

Also, adding the &quot;inline&quot; keyword to inlined JSValue methods causes a _ton_ of compiler errors.

It&apos;s not entirely clear how this ever worked; those inlined methods must have successfully been linked
against accidentally (which ironically defeats the purpose of inlining) by the linker so long as some
other cpp file in the same unified compilation file included the right Inlines.h header.

Another problem within JSC is that there are circular dependencies between headers; inlined methods
depend upon a type which is defined by a header that first includes the *Inlines.h header, causing
complicated compile errors in unrelated code.

To resolve all of these issues, all the externally defined inline functions in JSCJSValue.h are now
correctly attributed with &quot;inline&quot;, have comments declaring the header file containing the definition
of that inlined method, and in some cases the Inlines.h headers have been broken up into smaller
headers with fewer include requirements.

Canonical link: <a href="https://commits.webkit.org/302733@main">https://commits.webkit.org/302733@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/871830dca0d2da766e9cfc1927d7d535ee09d184

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2281 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40881 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137415 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81530 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d4cc8470-b720-4a49-972e-4786605569c0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131891 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2173 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99041 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/678a0c5f-0651-4486-94de-8667c9ed5fc9) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1664 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116454 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79741 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7c2de110-c238-437e-b68a-3d8f04db3afc) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1583 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34579 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80686 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122012 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110108 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35088 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139894 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128472 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2075 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1922 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107547 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112801 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107440 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27352 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1653 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/31256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54906 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2148 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65517 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161486 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1964 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40250 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1997 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2071 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.h:
```diff
@@ -460,25 +451,8 @@ class JSBigInt final : public JSCell {
-#endif
-        return static_cast<int64_t>(toBigUInt64Heap(bigInt.asHeapBigInt()));
-    }
+    inline static uint64_t toBigUInt64(JSValue); // Defined in JSBigIntInlines.h
+    inline static int64_t toBigInt64(JSValue); // Defined in JSBigIntInlines.h
```
Do add these comments elsewhere? It seems like `<file>Inlines.h` would be the first place I would look for the implementation.

jernoble replied:
Yes, it's common practice elsewhere in WebKit.  One benefit of this comment is that the correct header file to include shows up in the compiler error itself.

kmiller68 commented on Source/JavaScriptCore/runtime/JSCJSValueCellInlines.h:
```diff
@@ -0,0 +1,167 @@
+
+#pragma once
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
```
Why do you need these? Shouldn't `JSCJSValue.h` and/or `JSCell.h` annotate this correctly?

jernoble replied:
I only included them because they were present in the header file that these function definitions were moved from.

jernoble replied:
Okay, I'll remove these annotations; it doesn't appear to cause a build error.

kmiller68 commented on Source/JavaScriptCore/runtime/JSCJSValueCellInlines.h:
```diff
@@ -0,0 +1,167 @@
+
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+
+namespace JSC {
```
Probably worth a brief comment explaining why this file exists and that it's meant to break cycles between `JSCell` and `JSValue`.

kmiller68 reviewed (approved):
r=me with some comments.

jernoble committed:
[JSC] Mark inlined JSValue methods with `inline` keyword (and fix a bunch of things that break when that happens)
rdar://164200123
https://bugs.webkit.org/show_bug.cgi?id=302121

Reviewed by Keith Miller.

Clients of JSC can face "missing symbol" linker errors for inlined methods when the
correct *Inlines.h headers are not included, and that task is made much, much more difficult when
the inlined methods aren't marked as "inline". Adding the "inline" keyword causes the compiler to
throw an error when that symbol is referenced, making it much easier to notice a missing *Inlines.h
header at compile time.

Also, adding the "inline" keyword to inlined JSValue methods causes a _ton_ of compiler errors.

It's not entirely clear how this ever worked; those inlined methods must have successfully been linked
against accidentally (which ironically defeats the purpose of inlining) by the linker so long as some
other cpp file in the same unified compilation file included the right Inlines.h header.

Another problem within JSC is that there are circular dependencies between headers; inlined methods
depend upon a type which is defined by a header that first includes the *Inlines.h header, causing
complicated compile errors in unrelated code.

To resolve all of these issues, all the externally defined inline functions in JSCJSValue.h are now
correctly attributed with "inline", have comments declaring the header file containing the definition
of that inlined method, and in some cases the Inlines.h headers have been broken up into smaller
headers with fewer include requirements.

Canonical link: https://commits.webkit.org/302733@main
